File: service/game/actions/campfire.py
from service.game.actions.base import Command


class StartFireCommand(Command):
    def execute(self, game_state, player):
        if game_state.phase != "NIGHT" or player.fire_status == "HOST":
            return False

        cost = game_state.campfire_cost.get("wood", 1)
        if self._deduct_resources(player, {"wood": cost}):
            player.fire_status = "HOST"
            player.fire_guests = []
            player.add_timeline_event(
                "ACTION_COMPLETED", {"action": "START_FIRE"}
            )
            return True
        return False


def seat_guest(game_state, contract):
    if getattr(contract, "is_request", False):
        host = game_state.players.get(contract.target_id)
        guest = game_state.players.get(contract.initiator_id)
    else:
        host = game_state.players.get(contract.initiator_id)
        guest = game_state.players.get(contract.target_id)

    if not host or not guest:
        return False
    if guest.fire_status == "GUEST":
        return False
    if getattr(host, "fire_status", "COLD") != "HOST":
        return False
    if len(getattr(host, "fire_guests", [])) >= game_state.max_fire_seats:
        return False

    host.fire_guests.append(guest.session_id)
    guest.fire_status = "GUEST"
    guests = list(host.fire_guests)
    guest.fire_history.append({
        "host_id": host.session_id,
        "fire_id": contract.id,
        "role": "guest",
        "guests": guests.copy(),
    })
    host.fire_history.append({
        "host_id": host.session_id,
        "fire_id": contract.id,
        "role": "host",
        "guests": guests.copy(),
    })
    host.add_timeline_event("SEATED_GUEST", {"guest": guest.session_id})
    guest.add_timeline_event("JOINED_FIRE", {"host": host.session_id})
    return True

File: service/src/game/actions/campfire.ts
import {
  campfireActionSchema,
  type CampfireActionDTO,
  type FireHistoryDTO,
  type PlayerDTO,
} from "@takes-a-village/shared";
import { randomUUID } from "node:crypto";

interface CampfirePlayer {
  readonly id: string;
  readonly actions: PlayerDTO["actions"];
  readonly fireGuests: string[];
  readonly fireHistory: FireHistoryDTO[];
  fireStatus: PlayerDTO["fire_status"];
}

interface CampfireGame {
  readonly players: Map<string, CampfirePlayer>;
  readonly rules: { readonly MAX_FIRE_SEATS: number };
}

export function handleCampfireAction(
  game: CampfireGame,
  player: CampfirePlayer,
  command: unknown,
  payload: Record<string, unknown>,
): boolean {
  if (command === "CAMPFIRE") return draftCampfire(game, player, payload);
  if (command === "ACCEPT") return acceptCampfire(game, player, payload.action_id);
  return false;
}

export function cleanupCampfireActions(game: CampfireGame): void {
  for (const player of game.players.values()) {
    for (let index = player.actions.length - 1; index >= 0; index -= 1) {
      if (player.actions[index]?.type === "CAMPFIRE") player.actions.splice(index, 1);
    }
  }
}

function draftCampfire(
  game: CampfireGame,
  initiator: CampfirePlayer,
  payload: Record<string, unknown>,
): boolean {
  if (typeof payload.target_id !== "string") return false;
  const target = game.players.get(payload.target_id);
  const isRequest = payload.is_request === true;
  if (!target || (!isRequest && initiator.fireStatus !== "HOST")) return false;
  const parsed = campfireActionSchema.safeParse({
    id: randomUUID(),
    initiator_id: initiator.id,
    target_id: target.id,
    type: "CAMPFIRE",
    status: "PENDING",
    waiting_on_id: target.id,
    is_request: isRequest,
  });
  if (!parsed.success) return false;
  initiator.actions.push(parsed.data);
  target.actions.push(parsed.data);
  return true;
}

function acceptCampfire(game: CampfireGame, actor: CampfirePlayer, actionId: unknown): boolean {
  const contract = findCampfire(game, actionId);
  if (!contract || contract.status !== "PENDING" || contract.waiting_on_id !== actor.id) return false;
  const hostId = contract.is_request ? contract.target_id : contract.initiator_id;
  const guestId = contract.is_request ? contract.initiator_id : contract.target_id;
  const host = hostId ? game.players.get(hostId) : undefined;
  const guest = guestId ? game.players.get(guestId) : undefined;
  if (!host || !guest
    || host.fireStatus !== "HOST"
    || guest.fireStatus === "GUEST"
    || host.fireGuests.length >= game.rules.MAX_FIRE_SEATS) return false;

  host.fireGuests.push(guest.id);
  guest.fireStatus = "GUEST";
  const guests = [...host.fireGuests];
  host.fireHistory.push({ fire_id: contract.id, host_id: host.id, role: "host", guests });
  guest.fireHistory.push({ fire_id: contract.id, host_id: host.id, role: "guest", guests: [...guests] });
  contract.status = "ACCEPTED";
  contract.waiting_on_id = null;
  return true;
}

function findCampfire(game: CampfireGame, actionId: unknown): CampfireActionDTO | undefined {
  if (typeof actionId !== "string") return undefined;
  for (const player of game.players.values()) {
    const action = player.actions.find((candidate) => candidate.id === actionId);
    if (action?.type === "CAMPFIRE") return action as CampfireActionDTO;
  }
  return undefined;
}
